### ISSUE TITLE
Cannot read property 'set' of undefined

### ISSUE THREAD

**Issue by Jarrio** [Aidan63/ecs#3]
I keep getting this error in a basic project with a regular `setComponents(universe.createEntity(), ....)` where should I be looking to figure out this issue? I'm not sure if this is a bug or if i'm doing something wrong
![image](https://user-images.githubusercontent.com/748557/101986431-b7fc1500-3c85-11eb-94f0-1ad39dfc0dd2.png)

Id does have a value and the universe is instantiated in Main

**Comment by Jarrio** [comment]
added a trace for `components` on line 49 and the error changes to the following
![image](https://user-images.githubusercontent.com/748557/101986668-1aa1e080-3c87-11eb-8b57-c15e71667e51.png)

**Comment by Aidan63** [comment]
Haven't seen this before, do you have a small code sample which can trigger this error? Also if you try and run it through something like the interp target (if you can) what errors do you get then, might help narrow things down.

**Comment by Jarrio** [comment]
The code I was using was:
```haxe
package systems;

import systems.FireSystem;
import types.LogTypes;

class Logger extends FireSystem {
	@:fastFamily var logs = {id:String, log:LogData};
	override function onAdded() {
		trace("test");
		var data = new LogData(LogLevel.LOW, LogEvent.NONE, 'testing');
		var entity = universe.createEntity();
		universe.setComponents(entity, "system", data);
	}

	override function update(_dt:Float) {
			iterate(logs, {
				trace(log);
				trace(log.level);
				trace(log.event);
				trace(log.content);
			});
	}
}
```
I actually did that and the error went away, I switched back to js and it is no longer here again as well. It seems intermittent

**Comment by Jarrio** [comment]
The below code seems to consistently cause the error
Main.hx
```hx
import ecs.Universe;

import haxe.Timer;
import types.LogTypes;
import systems.*;
using ecs.macros.UniverseMacros;

class Main {
	static function main() {
		final universe = new Universe(1000);
		universe.setSystems(Logger);
		universe.setSystems(Repro);


		new Timer(150).run = function() {
			universe.update(1);
		}
	}
}
```
LogTypes
```hx
class LogData {
	public var level:LogLevel;
	public var event:LogEvent;
	public var content:String;
	public function new(level:LogLevel, event:LogEvent, content:String) {
		this.level = level;
		this.event = event;
		this.content = content;
	}
}

@:forward enum abstract LogLevel(Int) {
	var INFO;
	var LOW;
	var MEDIUM;
	var HIGH;

	public var priority(get, never):Bool;
	private function get_priority() {
		return this > 1;
	}
}

@:forward enum abstract LogEvent(String) {
	var NONE;
}
```
Logger
```haxe
package systems;

import systems.FireSystem;
import types.LogTypes;

class Logger extends FireSystem {
	@:fastFamily var logs = {id:String, log:LogData};
	override function onAdded() {
		trace("test");
		var data = new LogData(LogLevel.LOW, LogEvent.NONE, 'testing');
		var entity = universe.createEntity();
		universe.setComponents(entity, "system", data);
	}

	override function update(_dt:Float) {
			iterate(logs, {
				trace(log);
				trace(log.level);
				trace(log.event);
				trace(log.content);
			});
	}
}
```

Repro
```haxe
package systems;

import haxe.Json;
import ecs.System;

class Repro extends System {
	@:fastFamily var strategies = {id: String, strategy: Strategy};

	override function onAdded() {
		var json = "{value: 'test'}";
		var data = [];
		for (i in 0...5) {
			data.push(Json.parse(json));
		}

		for (key => item in data) {
			universe.setComponents(universe.createEntity(), '$key', new Strategy(item));
		}
	}
	override function update(_dt:Float) {
		iterate(strategies, {
			trace(id);
			trace(strategy);
		});
	}
}

class Strategy {
	var data:TData;
	public function new(data:TData) {
		this.data = data;
	}
}

typedef TData = {
	var value:String;
}
```

**Comment by Aidan63** [comment]
Hmm, I've tried running that example and seems to run fine for me with interp and hxnodejs.

![image](https://user-images.githubusercontent.com/10246746/101991103-57c79c00-3ca2-11eb-9ade-2d35680f33e2.png)

Only thing I've noticed is that your Logger system extends FireSystem which you didn't post the code for so I use the standard System as a parent.

**Comment by Jarrio** [comment]
![image](https://user-images.githubusercontent.com/748557/101999684-75fbbf00-3cd7-11eb-8c45-169b6d1a371c.png)
this is all firesystem was, but even without it I have the same issue
![image](https://user-images.githubusercontent.com/748557/101999699-9deb2280-3cd7-11eb-859b-c2230edaadfa.png)
where `Logger.hx` is:
![image](https://user-images.githubusercontent.com/748557/101999709-afccc580-3cd7-11eb-9f71-b3f5b850de33.png)
I am testing on nodejs, but any dependencies on node at this point are removed.

**Comment by Aidan63** [comment]
Not sure if its the cause (might be different node version behaviours?) but the json string in the Repro class won't parse with how you have it.
![image](https://user-images.githubusercontent.com/10246746/102012708-e3ddd000-3d43-11eb-873c-cc68e9302dbf.png)

Changing it to `var json = '{"value": "test"}';` works for me. Not sure if thats the cause but making that change is needed for me to get it to run. Below is the classes you posted merged into one file and working for me.

```haxe
import haxe.Timer;
import ecs.System;
import ecs.Universe;
import ecs.macros.UniverseMacros;
import haxe.Json;

using ecs.macros.UniverseMacros;

function main() {
    final universe = new Universe(8);

	universe.setSystems(Logger);
	universe.setSystems(Repro);
	
    new Timer(150).run = () -> {
        universe.update(1);
    }
}


class Logger extends System {
	@:fastFamily var logs = {id:String, log:LogData};
	override function onAdded() {
		trace("test");
		var data = new LogData(LogLevel.LOW, LogEvent.NONE, 'testing');
		var entity = universe.createEntity();
		universe.setComponents(entity, "system", data);
	}

	override function update(_dt:Float) {
		iterate(logs, {
			trace(log);
			trace(log.level);
			trace(log.event);
			trace(log.content);
		});
	}
}

class Repro extends System {
	@:fastFamily var strategies = {id: String, strategy: Strategy};

	override function onAdded() {
		var json = '{"value": "test"}';
		var data = [];
		for (i in 0...5) {
			data.push(Json.parse(json));
		}

		for (key => item in data) {
			universe.setComponents(universe.createEntity(), '$key', new Strategy(item));
		}
	}
	override function update(_dt:Float) {
		iterate(strategies, {
			trace(id);
			trace(strategy);
		});
	}
}

class Strategy {
	var data:TData;
	public function new(data:TData) {
		this.data = data;
	}
}

typedef TData = {
	var value:String;
}

class LogData {
	public var level:LogLevel;
	public var event:LogEvent;
	public var content:String;
	public function new(level:LogLevel, event:LogEvent, content:String) {
		this.level = level;
		this.event = event;
		this.content = content;
	}
}

@:forward enum abstract LogLevel(Int) {
	var INFO;
	var LOW;
	var MEDIUM;
	var HIGH;

	public var priority(get, never):Bool;
	private function get_priority() {
		return this > 1;
	}
}

@:forward enum abstract LogEvent(String) {
	var NONE;
}
```

**Comment by Jarrio** [comment]
Maybe you can see something, I forgot to say? But i've now commented everything that makes this project unique. I am using `node 14.5.0`
![gif](https://user-images.githubusercontent.com/748557/102022830-38537080-3d81-11eb-958b-a775f08d274f.gif)

**Comment by Aidan63** [comment]
Figured it out! After copying the layout of things exactly as they are in that gif I got the error and it seems to be an issue with the wildcard imports in Main. I ran into this issue a couple of days ago with components, I haven't looked into the specifics or potential workarounds but it seems like the macro API doesn't like resolving types when they were wildcard imported.
For now chaning the `import systems.*;` in Main.hx with `import systems.Logger; import systems.Repro;` fixes it for me.
